lib/ruleHelper.js

```diff
@@ -76,15 +77,77 @@ RuleHelper.prototype = {
         } else if (expression.type === "CallExpression") {
             allowed = this.isAllowedCallExpression(expression.callee, escapeObject.methods || VALID_UNWRAPPERS);
         } else if (expression.type === "BinaryExpression") {
-            allowed = ((this.allowedExpression(expression.left, escapeObject))
-            && (this.allowedExpression(expression.right, escapeObject)));
+            allowed = ((this.allowedExpression(expression.left, escapeObject, details))
+            && (this.allowedExpression(expression.right, escapeObject, details)));
         } else if (expression.type === "TSAsExpression") {
             // TSAsExpressions contain the raw javascript value in 'expression'
-            allowed = this.allowedExpression(expression.expression, escapeObject);
+            allowed = this.allowedExpression(expression.expression, escapeObject, details);
         } else if (expression.type === "TypeCastExpression") {
-            allowed = this.allowedExpression(expression.expression, escapeObject);
+            allowed = this.allowedExpression(expression.expression, escapeObject, details);
         } else if (Array.isArray(expression)) {
-            allowed = expression.every((e) => this.allowedExpression(e, escapeObject));
+            allowed = expression.every((e) => this.allowedExpression(e, escapeObject, details));
+        } else if (expression.type === "Identifier") {
+            // find declared variables and see which are literals
+            const scope = this.context.getScope(expression);
+            const variableInfo = scope.set.get(expression.name);
+
+            // let's see if this variable is allowed, by tracing it back through the code
+            if (variableInfo) {
+                // look if the var was defined as allowable
+                let definedAsAllowed = false;
+                for (const def of variableInfo.defs) {
+                    if (def.node.type === "VariableDeclarator") {
+                        if ((def.kind !== "let") && (def.kind !== "const")) {
+                            // We do not allow for identifiers declared with "var", as they can be overridden in a
+                            // way that is hard for us to follow (e.g., assignments to globalThis[theirNameAsString]).
+                            continue;
+                        }
```

Comment:
> Do you have some particular case in mind where continuing to check other definition instead of "considering disallowed" would be necessary?

I'm mostly not entirely sure how good/bad eslint is with lexical scoping and was hoping we'd be more helpful if we kept looking. I totally sympathize with your idea of being more careful here. I will make error out early and a note about a potential future optimization.

I don't think we need to take care of a fantasy-declaration name, as long as we have an explicit allow-list of "const" and "let". This would be more problematic if we tried to enumerate badness.

Reply:
> I don't think we need to take care of a fantasy-declaration name, as long as we have an explicit allow-list of "const" and "let". This would be more problematic if we tried to enumerate badness.

yeah, that the additional test for a "fantasy-declaration" was definitely intended as optional also from my perspective 

(I mentioned it only in case we want a test that ensure we will not miss that case in a hypothetical future rewrite we "forget" about why we use an explicit allowlist and end up rewriting it the other way around, but it is absolutely true that with the current implementation that is going to be handled exactly like a "var").